crates/ruff_linter/src/rules/pylint/rules/unnecessary_list_index_lookup.rs

```diff
@@ -0,0 +1,275 @@
+                                if let Expr::Name(ast::ExprName { id, .. }) = slice.as_ref() {
+                                    if id == self.index_name {
+                                        return true;
+                                    }
+                                }
+                            }
+                        }
+                        false
+                    }
+                    _ => false,
+                });
+            }
+            _ => visitor::walk_stmt(self, stmt),
+        }
+    }
+}
+
+/// PLR1736
+pub(crate) fn unnecessary_list_index_lookup(checker: &mut Checker, stmt_for: &StmtFor) {
+    let Some((sequence, index_name, value_name)) =
+        enumerate_items(checker, &stmt_for.iter, &stmt_for.target)
+    else {
+        return;
+    };
+
+    let mut visitor = SubscriptVisitor::new(&sequence, &index_name);
+
+    visitor.visit_body(&stmt_for.body);
+    visitor.visit_body(&stmt_for.orelse);
+
+    for range in visitor.diagnostic_ranges {
+        let mut diagnostic = Diagnostic::new(UnnecessaryListIndexLookup, range);
+
+        diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+            value_name.clone(),
+            range,
+        )));
+
+        checker.diagnostics.push(diagnostic);
+    }
+}
+
+/// PLR1736
+pub(crate) fn unnecessary_list_index_lookup_comprehension(checker: &mut Checker, expr: &Expr) {
+    match expr {
+        Expr::GeneratorExp(ast::ExprGeneratorExp {
+            elt, generators, ..
+        })
+        | Expr::DictComp(ast::ExprDictComp {
+            value: elt,
+            generators,
+            ..
+        })
+        | Expr::SetComp(ast::ExprSetComp {
+            elt, generators, ..
+        })
+        | Expr::ListComp(ast::ExprListComp {
+            elt, generators, ..
+        }) => {
+            for comp in generators {
+                if let Some((sequence, index_name, value_name)) =
+                    enumerate_items(checker, &comp.iter, &comp.target)
+                {
+                    let mut visitor = SubscriptVisitor::new(&sequence, &index_name);
+
+                    visitor.visit_expr(elt.as_ref());
+
+                    for range in visitor.diagnostic_ranges {
+                        let mut diagnostic = Diagnostic::new(UnnecessaryListIndexLookup, range);
+
+                        diagnostic.set_fix(Fix::safe_edit(Edit::range_replacement(
+                            value_name.clone(),
+                            range,
+                        )));
+
+                        checker.diagnostics.push(diagnostic);
+                    }
+                }
+            }
+        }
+        _ => (),
+    }
+}
+
+fn enumerate_items(
+    checker: &mut Checker,
+    call_expr: &Expr,
+    tuple_expr: &Expr,
+) -> Option<(String, String, String)> {
+    let Expr::Call(ast::ExprCall {
+        func, arguments, ..
+    }) = call_expr
+    else {
+        return None;
+    };
```

Comment:
We could use the `call_expr.as_call_expr()?` pattern throughout this function. For example, the highlighted code would become:

```rust
let Some(ast::ExprCall { func, arguments, .. }) = call_expr.as_call_expr()?;
```

Reply:
Looks like the right code was

```rust
let ast::ExprCall {
    func, arguments, ..
} = call_expr.as_call_expr()?;
```

That tripped me up for a sec! 🤣 Thanks for the push in the right direction though!

Reply:
Ah right, sorry!